test/util.sh

```diff
@@ -0,0 +1,203 @@
+function os::test::expect_success_and_not_text() {
+	cmd=$1
+	expected_text=$2
+
+	os::test::expect_exit_code_run_grep "${cmd}" "success_func" "${expected_text}" "failure_func"
+}
+
+# expect_failure_and_not_text runs the cmd and expects a non-zero exit code
+# as well as running a grep test to ensure the given string is not in the output
+function os::test::expect_failure_and_not_text() {
+	cmd=$1
+	expected_text=$2
+
+	os::test::expect_exit_code_run_grep "${cmd}" "failure_func" "${expected_text}" "failure_func"
+}
+
+# MEDIUM-LEVEL WRAPPER FUNCTIONS: ASSUME NO CODES
+
+# expect_code runs the cmd and expects a given exit code
+function os::test::expect_code() {
+	cmd=$1
+	expected_cmd_code=$2
+
+	os::test::expect_exit_code_run_grep "${cmd}" "specific_code_func ${expected_cmd_code}"
+}
+
+# expect_code_and_text runs the cmd and expects the given exit code
+# as well as running a grep test to find the given string in the output
+function os::test::expect_code_and_text() {
+	cmd=$1
+	expected_cmd_code=$2
+	expected_text=$3
+
+	os::test::expect_exit_code_run_grep "${cmd}" "specific_code_func ${expected_cmd_code}" "${expected_text}"
+}
+
+# expect_code_and_not_text runs the cmd and expects the given exit code
+# as well as running a grep test to ensure the given string is not in the output
+function os::test::expect_code_and_not_text() {
+	cmd=$1
+	expected_cmd_code=$2
+	expected_text=$3
+
+	os::test::expect_exit_code_run_grep "${cmd}" "specific_code_func ${expected_cmd_code}" "${expected_text}" "failure_func"
+}
+
+# TEST EXECUTION FUNCTION
+
+# expect_exit_code_and_text runs the provided test command and expects a specific exit code from that
+# command as well as the success of a specified `grep` invocation. Output from the command to be 
+# tested is suppressed unless either `VERBOSE=1` or the test fails.
+function os::test::expect_exit_code_run_grep() {
+	cmd=$1
+	# default expected cmd code to 0 for success
+	cmd_eval_func=${2:-success_func}
+	# default grep args to a something that always passes, even for zero-length input
+	grep_args=${3:-'$'} 
+	# default expected test code to 0 for success
+	test_eval_func=${4:-success_func}
+
+	echo "Running  ${cmd}..."
+	
+	set +o errexit
+	cmd_output=$(${cmd} 2>&1)
+	cmd_result=$?
+	cmd_succeeded=$( ${cmd_eval_func} "${cmd_result}"; echo $? )
+
+	test_output=$(echo "${cmd_output}" | grep -E "${grep_args}" 2>&1)
+	test_result=$?
+	test_succeeded=$( ${test_eval_func} "${test_result}"; echo $? )
+	set -o errexit
+
+	if (( $cmd_succeeded && $test_succeeded )); then
+		overwrite_green "SUCCESS: ${cmd}"
+
+		if [[ -n ${VERBOSE-} ]]; then
+			print_results "${cmd_output}"
+		fi
+
+		return 0
+	else
+		cause=$(assemble_causes "${cmd_succeeded}" "${test_succeeded}")
+		
+		overwrite_red "FAILURE: ${cmd}: ${cause}"
+		print_results "${cmd_output}"
+
+		return 1
+	fi
+}
+
+# success_func determines if the input exit code denotes success
+function success_func() {
```

Comment:
I was thinking that it would be better not to namespace these as they are not really functions I would expect other people to be using. I will ask @Miciah if there is a good way to make it possible to import only the "public" functions. If not, should we still namespace these if we don't want to promote their use?